pennylane/templates/embeddings/qaoa.py

```diff
@@ -260,22 +200,136 @@ def circuit(weights, f=None):
         1-dimensional Ising model.
 
     """
-    wires = Wires(wires)
-    repeat = _preprocess(features, wires, weights)
-
-    if local_field == "Z":
-        local_fields = RZ
-    elif local_field == "X":
-        local_fields = RX
-    elif local_field == "Y":
-        local_fields = RY
-    else:
-        raise ValueError(f"did not recognize local field {local_field}")
 
-    for l in range(repeat):
-        # apply alternating Hamiltonians
-        qaoa_feature_encoding_hamiltonian(features, wires)
-        qaoa_ising_hamiltonian(weights[l], wires, local_fields)
+    num_params = 2
+    num_wires = AnyWires
+    par_domain = "A"
+
+    def __init__(self, features, weights, wires, local_field="Y", do_queue=True):
+
+        if local_field == "Z":
+            self.local_field = qml.RZ
+        elif local_field == "X":
+            self.local_field = qml.RX
+        elif local_field == "Y":
+            self.local_field = qml.RY
+        else:
+            raise ValueError(f"did not recognize local field {local_field}")
+
+        super().__init__(features, weights, wires=wires, do_queue=do_queue)
+        self._preprocess()
+
+    def expand(self):
+
+        features = self.data[0]
+        weights = self.data[1]
+
+        # first dimension of the weights tensor determines
+        # the number of layers
+        repeat = qml.math.shape(weights)[0]
+
+        with qml.tape.QuantumTape() as tape:
+
+            for l in range(repeat):
```

Comment:
Current implementation is fine, but for the future, the [`qml.layer()`](https://pennylane.readthedocs.io/en/latest/code/api/pennylane.layer.html?highlight=layer#pennylane.layer) function could provide perhaps a cleaner implementation

Reply:
That's the point above - we found that `broadcast` and `layer` have quite a large overhead, because they have to cater for general inputs and  check their dimensions. So we should try to always code up the decomposion in a vanilla fashion!

Reply:
flyby comment: I worry `qml.layer()` may suffer from the same slow down as `qml.broadcast`, so this might be intentional on @mariaschuld's part

Reply:
Oh Josh already said that :)

Reply:
I think we both commented at the exact same time, oops.